<h1>Loading Necessary Libraries</h1>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

<h2>Data PreProcessing Starts</h2>

In [2]:
df=pd.read_csv('test.csv')
df_survived=pd.read_csv('gender_submission.csv')

In [3]:
df.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [4]:
df_survived.head()

PassengerId  Survived
0          892         0
1          893         1
2          894         0
3          895         0
4          896         1

# Concating df and df_survived because df doesnot have survived column and both are aligned perfectly


In [5]:
df=pd.concat([df,df_survived['Survived']],axis=1)

In [6]:
df.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  Survived  
0  34.5      0      0   330911   7.8292   NaN        Q         0  
1  47.0      1      0   363272   7.0000   NaN        S         1  
2  62.0      0      0   240276   9.6875   NaN        Q         0  
3  27.0      0      0   315154   8.6625   NaN        S         0  
4  22.0      1      1  3101298  12.2875   NaN        S         1

In [7]:
df.sample(5)

PassengerId  Pclass                                             Name  \
162         1054       2                             Wright, Miss. Marion   
223         1115       3                    Karlsson, Mr. Einar Gervasius   
119         1011       2  Chapman, Mrs. John Henry (Sara Elizabeth Lawry)   
253         1145       3                         Salander, Mr. Karl Johan   
29           921       3                                Samaan, Mr. Elias   

        Sex   Age  SibSp  Parch       Ticket     Fare Cabin Embarked  Survived  
162  female  26.0      0      0       220844  13.5000   NaN        S         1  
223    male  21.0      0      0       350053   7.7958   NaN        S         0  
119  female  29.0      1      0  SC/AH 29037  26.0000   NaN        S         1  
253    male  24.0      0      0         7266   9.3250   NaN        S         0  
29     male   NaN      2      0         2662  21.6792   NaN        C         0

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
 11  Survived     418 non-null    int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 39.3+ KB


In [9]:
df.shape

(418, 12)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
 11  Survived     418 non-null    int64  
dtypes: float64(2), int64(5), object(5)
memory usage: 39.3+ KB


# We have 418 total datas and in some we can observe value for nulls. 
Checking for null values for each columns

In [11]:
df.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
Survived         0
dtype: int64

# Determining how to fill datas for age


In [12]:
print('Oldest Passenger was of:',df['Age'].max(),'Years')
print('Youngest Passenger was of:',df['Age'].min(),'Years')
print('Average Age on the ship:',df['Age'].mean(),'Years')

#So maybe mean would be appropriate

Oldest Passenger was of: 76.0 Years
Youngest Passenger was of: 0.17 Years
Average Age on the ship: 30.272590361445783 Years



# Determining how to fill datas for Cabins


In [13]:
print(len(df['Cabin'].unique()))

#Many unique Values. 

print(df['Cabin'].sample(10))

#Values seem not appropiate to take mean.


77
123                NaN
64     B57 B59 B63 B66
188                NaN
379                NaN
208                A11
122                C28
4                  NaN
52                 NaN
191                NaN
350            D10 D12
Name: Cabin, dtype: object


In [14]:

# Checking Cabin's Relevance to output


In [15]:

#df['Cabin'].corr(df['Survived']) Couldnot be computed

#Also if we analyze Cabin will have no importance to survive rate

#Best to drop them.

# Determining how to fill datas for Fare


In [16]:
print(len(df['Fare'].unique()))

#Has 170 unique Values. 
#So a Mean would be appropiate


170


## Lets Fill NAN Feilds

In [17]:
df['Fare'].fillna(df['Fare'].mean(),inplace=True) #Replacing with mean values
df['Age'].fillna(df['Age'].mean(),inplace=True) #Replacing with mean values
df.drop('Cabin',axis=1,inplace=True)

In [18]:
df.columns

Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Embarked', 'Survived'],
      dtype='object')

Age has 177 null values
Cabin has 687 null values
Emabrked has 2 null values

We can drop them. But 177 and 687 are too large numbers to drop.
Further preprocessing might be required.

<h1> Now Lets Look at Feature Selection
 </h1>

In [19]:
#Looking at correlation between columns with survival

df_comp=df.iloc[:,df.columns!='Survived']
df_comp.corrwith(df['Survived'])


PassengerId   -0.023245
Pclass        -0.108615
Age           -0.000011
SibSp          0.099943
Parch          0.159120
Fare           0.191382
dtype: float64

In [20]:
#dropping PassengerID , Age , SibSp 
rows=['PassengerId','Age','SibSp']
for r in rows:
    df.drop(r,axis=1,inplace=True)

In [21]:
df.columns

Index(['Pclass', 'Name', 'Sex', 'Parch', 'Ticket', 'Fare', 'Embarked',
       'Survived'],
      dtype='object')

# Checking realations for<br>
## <b>Name  , Ticket And  Sex</b>


#We cannont find Correlation between Name and Survived

#Name Cannot be Categorized 

#Possibly Survival rate cannot be determined by only names.

#So better drop it


In [22]:
df.drop('Name',axis=1,inplace=True)

In [23]:
#Tickets

df['Ticket'].sample(10)

304                383123
292                  2652
46               PC 17594
68                   2543
218                113503
240                112377
361       S.C./PARIS 2079
325    SOTON/O.Q. 3101309
249                  2660
65           F.C.C. 13534
Name: Ticket, dtype: object

# Tickets some have String values and they all seem to be random.
#So lets drop it as ticket.


In [24]:
df.drop('Ticket',axis=1,inplace=True)


In [25]:
# For Sex
df['Sex'].unique()

array(['male', 'female'], dtype=object)

# Lets map Male and Female to 0s and 1s so as to compare correlation


In [26]:
df['Sex']=df['Sex'].apply(lambda x:1 if x=='female' else 0)

# Lets Check correlation


In [27]:
df['Sex'].value_counts()
df['Sex'].corr(df['Survived'])

0.9999999999999999

Thats very high Correlation. So lets keep Sex as feature


In [28]:

df.columns.values

array(['Pclass', 'Sex', 'Parch', 'Fare', 'Embarked', 'Survived'],
      dtype=object)

# column names


In [29]:
col_names=df.columns.values 

for i in range(0,len(col_names)):
    uniq_datas=df[col_names[i]].unique() #unique datas of a column 
    
    if(len( uniq_datas ) < 10 ):
        print(col_names[i] ," : ", np.sort(uniq_datas))
    
    else:
        arr= df[col_names[i]].sort_values()
        chunks = len(df[col_names[i]])/100
        print( col_names[i]," : " ,np.array_split(arr,chunks)) #splitting in chunks
    

Pclass  :  [1 2 3]
Sex  :  [0 1]
Parch  :  [0 1 2 3 4 5 6 9]
Fare  :  [372    0.0000
266    0.0000
21     3.1708
133    6.4375
116    6.4375
        ...  
86     7.8792
381    7.8792
243    7.8875
60     7.8958
134    7.8958
Name: Fare, Length: 105, dtype: float64, 295     7.8958
334     7.8958
56      7.8958
71      7.8958
209     7.8958
        ...   
309    14.1083
201    14.4000
85     14.4542
249    14.4542
115    14.4542
Name: Fare, Length: 105, dtype: float64, 366    14.4583
382    14.5000
25     14.5000
344    14.5000
192    14.5000
        ...   
379    31.3875
174    31.3875
154    31.3875
338    31.5000
31     31.5000
Name: Fare, Length: 104, dtype: float64, 319     31.5000
112     31.6792
22      31.6833
194     32.5000
247     32.5000
         ...   
375    262.3750
24     262.3750
53     263.0000
69     263.0000
343    512.3292
Name: Fare, Length: 104, dtype: float64]
Embarked  :  ['C' 'Q' 'S']
Survived  :  [0 1]


# Now converting Categorical data Embarked into codes


In [30]:

embarked_df=pd.get_dummies(df['Embarked'])

df=pd.concat([df,embarked_df],axis=1)

df.drop('Embarked',axis=1,inplace=True)


# Now converting Categorical data Parch into codes


In [31]:
parched_df=pd.get_dummies(df['Parch'])

df=pd.concat([df,parched_df],axis=1)

df.drop('Parch',axis=1,inplace=True)


In [32]:
df.head()

Pclass  Sex     Fare  Survived  C  Q  S  0  1  2  3  4  5  6  9
0       3    0   7.8292         0  0  1  0  1  0  0  0  0  0  0  0
1       3    1   7.0000         1  0  0  1  1  0  0  0  0  0  0  0
2       2    0   9.6875         0  0  1  0  1  0  0  0  0  0  0  0
3       3    0   8.6625         0  0  0  1  1  0  0  0  0  0  0  0
4       3    1  12.2875         1  0  0  1  0  1  0  0  0  0  0  0

# Features Cannot be Int so Changing The Feature Names

In [33]:
for i in range(0,10):
    df.rename(columns={i:f"Parch_{i}"},inplace=True)

In [34]:
df.head()

Pclass  Sex     Fare  Survived  C  Q  S  Parch_0  Parch_1  Parch_2  \
0       3    0   7.8292         0  0  1  0        1        0        0   
1       3    1   7.0000         1  0  0  1        1        0        0   
2       2    0   9.6875         0  0  1  0        1        0        0   
3       3    0   8.6625         0  0  0  1        1        0        0   
4       3    1  12.2875         1  0  0  1        0        1        0   

   Parch_3  Parch_4  Parch_5  Parch_6  Parch_9  
0        0        0        0        0        0  
1        0        0        0        0        0  
2        0        0        0        0        0  
3        0        0        0        0        0  
4        0        0        0        0        0

In [35]:
#Checking correlations
df.corr()

Pclass       Sex      Fare  Survived         C         Q  \
Pclass    1.000000 -0.108615 -0.576619 -0.108615 -0.318543  0.252608   
Sex      -0.108615  1.000000  0.191382  1.000000  0.033684  0.115574   
Fare     -0.576619  0.191382  1.000000  0.191382  0.312040 -0.155539   
Survived -0.108615  1.000000  0.191382  1.000000  0.033684  0.115574   
C        -0.318543  0.033684  0.312040  0.033684  1.000000 -0.199786   
Q         0.252608  0.115574 -0.155539  0.115574 -0.199786  1.000000   
S         0.120783 -0.105883 -0.178464 -0.105883 -0.767375 -0.474962   
Parch_0   0.054258 -0.224158 -0.256771 -0.224158 -0.080877  0.171098   
Parch_1  -0.093179  0.136987  0.148275  0.136987  0.140263 -0.109384   
Parch_2   0.034153  0.147542  0.096940  0.147542 -0.042399 -0.102952   
Parch_3  -0.094264  0.053559  0.232155  0.053559  0.083666 -0.029898   
Parch_4  -0.021898  0.091725  0.135019  0.091725 -0.039394 -0.024382   
Parch_5   0.042775 -0.037018 -0.003723 -0.037018 -0.027822 -0.017220   
Parch_6   0.042775 -0.037018  0.009898 -0.037018 -0.027822 -0.017220   
Parch_9   0.060565  0.019655  0.042173  0.019655 -0.039394 -0.024382   

                 S   Parch_0   Parch_1   Parch_2   Parch_3   Parch_4  \
Pclass    0.120783  0.054258 -0.093179  0.034153 -0.094264 -0.021898   
Sex      -0.105883 -0.224158  0.136987  0.147542  0.053559  0.091725   
Fare     -0.178464 -0.256771  0.148275  0.096940  0.232155  0.135019   
Survived -0.105883 -0.224158  0.136987  0.147542  0.053559  0.091725   
C        -0.767375 -0.080877  0.140263 -0.042399  0.083666 -0.039394   
Q        -0.474962  0.171098 -0.109384 -0.102952 -0.029898 -0.024382   
S         1.000000 -0.039328 -0.054392  0.105450 -0.055576  0.051335   
Parch_0  -0.039328  1.000000 -0.699793 -0.543545 -0.157850 -0.128729   
Parch_1  -0.054392 -0.699793  1.000000 -0.110354 -0.032048 -0.026135   
Parch_2   0.105450 -0.543545 -0.110354  1.000000 -0.024892 -0.020300   
Parch_3  -0.055576 -0.157850 -0.032048 -0.024892  1.000000 -0.005895   
Parch_4   0.051335 -0.128729 -0.026135 -0.020300 -0.005895  1.000000   
Parch_5   0.036256 -0.090916 -0.018458 -0.014337 -0.004164 -0.003395   
Parch_6   0.036256 -0.090916 -0.018458 -0.014337 -0.004164 -0.003395   
Parch_9   0.051335 -0.128729 -0.026135 -0.020300 -0.005895 -0.004808   

           Parch_5   Parch_6   Parch_9  
Pclass    0.042775  0.042775  0.060565  
Sex      -0.037018 -0.037018  0.019655  
Fare     -0.003723  0.009898  0.042173  
Survived -0.037018 -0.037018  0.019655  
C        -0.027822 -0.027822 -0.039394  
Q        -0.017220 -0.017220 -0.024382  
S         0.036256  0.036256  0.051335  
Parch_0  -0.090916 -0.090916 -0.128729  
Parch_1  -0.018458 -0.018458 -0.026135  
Parch_2  -0.014337 -0.014337 -0.020300  
Parch_3  -0.004164 -0.004164 -0.005895  
Parch_4  -0.003395 -0.003395 -0.004808  
Parch_5   1.000000 -0.002398 -0.003395  
Parch_6  -0.002398  1.000000 -0.003395  
Parch_9  -0.003395 -0.003395  1.000000

# Here Lets look between Pclass and Fare


In [36]:

df['Pclass'].corr(df['Fare'])

-0.5766190772528164

# A close correlation. Maybe we can discard a feature.
# checking between Survived


In [37]:

print("With Pclass")
print(df['Pclass'].corr(df['Survived']))

print("With Fare")
print(df['Fare'].corr(df['Survived']))

With Pclass
-0.10861452091523417
With Fare
0.19138247897510113


# Closer relation with Fare so lets drop PClass


In [38]:
df.drop('Pclass',axis=1,inplace=True)

In [39]:
df.head()

Sex     Fare  Survived  C  Q  S  Parch_0  Parch_1  Parch_2  Parch_3  \
0    0   7.8292         0  0  1  0        1        0        0        0   
1    1   7.0000         1  0  0  1        1        0        0        0   
2    0   9.6875         0  0  1  0        1        0        0        0   
3    0   8.6625         0  0  0  1        1        0        0        0   
4    1  12.2875         1  0  0  1        0        1        0        0   

   Parch_4  Parch_5  Parch_6  Parch_9  
0        0        0        0        0  
1        0        0        0        0  
2        0        0        0        0  
3        0        0        0        0  
4        0        0        0        0

<h1> Now Build Model </h1>

In [40]:
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score
from sklearn.model_selection import GridSearchCV

In [41]:
#Now Finally Implementing Machine Learning
svc=SVC()
knn=KNeighborsClassifier()
dtc=DecisionTreeClassifier()
nb=GaussianNB()

In [42]:
models={'svc':svc,'knn':knn,'dtc':dtc,'nb':nb}

In [43]:
x,y=df.drop('Survived',axis=1),df['Survived']

In [44]:
train_x,test_x,train_y,test_y=train_test_split(x,y,test_size=0.2,random_state=42)

In [45]:
model_accuracy={}

for model in models.keys():
    train_x,test_x,train_y,test_y=train_test_split(x,y,test_size=0.2,random_state=500)
    mod = models[model].fit(train_x,train_y)
    predict=mod.predict(test_x)
    #print(f"{model}",classification_report(predict,test_y))
    print(f"{model}")
    print(confusion_matrix(predict,test_y))
    
    if model not in model_accuracy.keys():
        model_accuracy.update({model:accuracy_score(test_y,predict)})

svc
[[55 27]
 [ 1  1]]
knn
[[54  5]
 [ 2 23]]
dtc
[[56  0]
 [ 0 28]]
nb
[[56  0]
 [ 0 28]]


In [46]:
model_accuracy

{'svc': 0.6666666666666666, 'knn': 0.9166666666666666, 'dtc': 1.0, 'nb': 1.0}

# Lets try With GridSearch


In [47]:
params={
    'knn': {'n_neighbors':[3,5,7,9,11,13,15],
    'metric':['cosine','euclidean','manhattan'],
    'weights':['uniform','distance']},
    
    'svc': {
    'C':[0.1,1,10,100], 
    'gamma':[1,0.1,0.01,0.01], 
    'kernel':['rbf','linear']},
    
    'dtc':{
        'criterion':['gini','entropy'],
        'max_depth':[2,4,6,8,10,12]
    },
    
    'nb':{
        'priors': [None],
    'var_smoothing': [0.00000001, 0.000000001, 0.00000001]
    }
    
    }

In [50]:
model_accuracy={}
for model in models.keys():
    mod = GridSearchCV(
    models[model],
    params[model],
    verbose=1, #Progress bar showing
    cv=20, #cross validation
    n_jobs=-1, #cores to assign
)
    gridsearch_result=mod.fit(train_x,train_y)
    predict=mod.predict(test_x)
    print(f"{model} : ",gridsearch_result.best_estimator_)
    print(f"{model}")
    print(confusion_matrix(predict,test_y))
    if model not in model_accuracy.keys():
        model_accuracy.update({model:accuracy_score(test_y,predict)})

Fitting 20 folds for each of 32 candidates, totalling 640 fits
svc :  SVC(C=0.1, gamma=1, kernel='linear')
svc
[[56  0]
 [ 0 28]]
Fitting 20 folds for each of 42 candidates, totalling 840 fits
knn :  KNeighborsClassifier(metric='cosine', n_neighbors=3, weights='distance')
knn
[[56  1]
 [ 0 27]]
Fitting 20 folds for each of 12 candidates, totalling 240 fits
dtc :  DecisionTreeClassifier(max_depth=2)
dtc
[[56  0]
 [ 0 28]]
Fitting 20 folds for each of 3 candidates, totalling 60 fits
nb :  GaussianNB(var_smoothing=1e-08)
nb
[[56  0]
 [ 0 28]]


In [51]:
model_accuracy

{'svc': 1.0, 'knn': 0.9880952380952381, 'dtc': 1.0, 'nb': 1.0}

In [ ]:
#Maybe the result is nice.